File: app/src/main/java/com/quranapp/android/components/quran/subcomponents/Verse.java
package com.quranapp.android.components.quran.subcomponents;

import android.content.Context;
import androidx.annotation.NonNull;

import com.quranapp.android.utils.verse.VerseUtils;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class Verse implements Serializable {
    public final int id;
    public final int pageNo;
    public final int chapterNo;
    public final int verseNo;
    public final String arabicText;
    private List<Translation> translations = new ArrayList<>();
    private boolean includeChapterNameInSerial;
    private transient CharSequence mTranslTextSpannable;

    public Verse(int id, int chapterNo, int verseNo, int pageNo, String arabicText) {
        this.id = id;
        this.chapterNo = chapterNo;
        this.verseNo = verseNo;
        this.pageNo = pageNo;
        this.arabicText = arabicText;
    }

    public Verse(Verse verse) {
        id = verse.id;
        chapterNo = verse.chapterNo;
        verseNo = verse.verseNo;
        pageNo = verse.pageNo;
        arabicText = verse.arabicText;
        // not copying
        translations = verse.translations;
        includeChapterNameInSerial = verse.includeChapterNameInSerial;
    }

    public List<Translation> getTranslations() {
        return translations;
    }

    public void setTranslations(List<Translation> translations) {
        this.translations = translations;
    }

    public int getTranslationCount() {
        return translations.size();
    }

    public CharSequence getTranslTextSpannable() {
        return mTranslTextSpannable;
    }

    public void setTranslTextSpannable(CharSequence translationTextSpannable) {
        mTranslTextSpannable = translationTextSpannable;
    }

    public boolean getIncludeChapterNameInSerial() {
        return includeChapterNameInSerial;
    }

    public void setIncludeChapterNameInSerial(boolean include) {
        includeChapterNameInSerial = include;
    }

    public boolean isVOTD(Context ctx) {
        return VerseUtils.isVOTD(ctx, chapterNo, verseNo);
    }

    public boolean isIdealForVOTD() {
        int l = arabicText.length();
        return l > 5 && l <= 300;
    }

    public Verse copy() {
        return new Verse(this);
    }

    @NonNull
    @Override
    public String toString() {
        return String.format(Locale.getDefault(), "VERSE: ChapterNo - %d, VerseNo - %d\n", chapterNo, verseNo);
    }
}

File: app/src/main/java/com/quranapp/android/components/quran/subcomponents/Verse.kt
package com.quranapp.android.components.quran.subcomponents

import android.content.Context
import androidx.compose.ui.text.AnnotatedString
import com.quranapp.android.utils.verse.VerseUtils
import java.io.Serializable
import java.util.Locale


class Verse : Serializable {
    val id: Int

    @JvmField
    val pageNo: Int

    @JvmField
    val chapterNo: Int

    @JvmField
    val verseNo: Int

    @JvmField
    val arabicText: String = ""

    var segments: List<String> = emptyList()

    @JvmField
    val endText: String
    var translations: List<Translation> = ArrayList()
    var includeChapterNameInSerial = false

    @Deprecated("")
    @Transient
    @JvmField
    var arabicTextSpannable: CharSequence? = null

    @Deprecated("")
    @Transient
    @JvmField
    var translTextSpannable: CharSequence? = null

    constructor(
        id: Int,
        chapterNo: Int,
        verseNo: Int,
        pageNo: Int,
        segments: List<String>,
        endText: String,
    ) {
        this.id = id
        this.chapterNo = chapterNo
        this.verseNo = verseNo
        this.pageNo = pageNo
        this.segments = segments
        this.endText = endText
    }

    constructor(verse: Verse) {
        id = verse.id
        chapterNo = verse.chapterNo
        verseNo = verse.verseNo
        pageNo = verse.pageNo
        segments = verse.segments
        endText = verse.endText

        // not copying
        translations = verse.translations
        includeChapterNameInSerial = verse.includeChapterNameInSerial
    }

    fun getTranslationCount() = translations.size

    fun isVOTD(ctx: Context) = VerseUtils.isVOTD(chapterNo, verseNo)

    fun isIdealForVOTD() = arabicText.length in 6..300

    fun copy() = Verse(this)

    override fun toString(): String {
        return String.format(
            Locale.getDefault(),
            "VERSE: ChapterNo - %d, VerseNo - %d\n",
            chapterNo,
            verseNo
        )
    }
}